source/core/icosahedron.ts

```diff
@@ -0,0 +1,212 @@
+
+import { vec2, vec3 } from 'gl-matrix';
+import { v3 } from '../gl-matrix-extensions';
+
+type ivec3 = [number, number, number];
+
+/**
+ * Helper class to generate a sphere geometry based on a refinable icosahedron
+ */
+export class Icosahedron {
+
+    protected _faces: Array<ivec3>;     // List of faces (triangles)
+    protected _vertices: Array<vec3>;  // List of vertices
+    protected _texcoords: Array<vec2>; // List of texture coordinates
+
+    /**
+     * Object constructor.
+     */
+    constructor() {
+    }
+
+    /**
+     * Returns the base vertices for the icosahedron.
+     */
+    protected baseVertices(): Array<vec3> {
+        // Icosahedron vertices (normalized) form three orthogonal golden rectangles:
+        // http://en.wikipedia.org/wiki/Icosahedron#Cartesian_coordinates
+
+        const t = (1.0 + Math.sqrt(5.0)) * 0.5;
+        const i = 1.0 / Math.sqrt(t * t + 1.0);
+        const a = t * i;
+
+        const vertices: Array<vec3> = [
+            vec3.fromValues(-i, a, 0),
+            vec3.fromValues(i, a, 0),
+            vec3.fromValues(-i, -a, 0),
+            vec3.fromValues(i, -a, 0),
+            vec3.fromValues(0, -i, a),
+            vec3.fromValues(0, i, a),
+            vec3.fromValues(0, -i, -a),
+            vec3.fromValues(0, i, -a),
+            vec3.fromValues(a, 0, -i),
+            vec3.fromValues(a, 0, i),
+            vec3.fromValues(-a, 0, -i),
+            vec3.fromValues(-a, 0, i),
+        ];
+
+        return vertices;
+    }
+
+    /**
+     * Returns the base faces for the icosahedron.
+     */
+    protected baseFaces(): Array<ivec3> {
+        const faces: Array<ivec3> = [
```

Comment:
same as for the box, we should extract this to geometry data files and then just use transpile-time include